### PR TITLE
initialise_aj

### LINKED ISSUE #11: Add a lecture on the Aiyagari model
This was started by @jstac in https://github.com/QuantEcon/cbc_workshops/tree/main/aiyagari and @shlff has added some fixes.

@shlff, please collaborate with the team on this one.

### PR DESCRIPTION
**Pull request by shlff**
This fix #11 .

### CODE REVIEW

**Comment by shlff** [comment]
To-do:
- [ ] add static argnum flags
- [ ] add exercise 1
- [ ] add exercise 2

Exercise 1 asks the reader to plot how the equilibrium changes as beta increases.

Exercise 2 asks the reader to change agent's utility function from $u(c) = log c$ to $u(c) = c^{1-\gamma} / (1-\gamma)$ where $\gamma =2$ and then recompute the equilibrium.

File: lectures/aiyagari_jax.md [hunk 1]
```diff
@@ -0,0 +1,925 @@
...
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.14.4
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
++++ {"colab_type": "text", "id": "view-in-github"}
```
Comment (jstac) [inline lectures/aiyagari_jax.md:14]: @shlff please remove all this meta data inserted by colab --- it should  not be in our md files

File: lectures/aiyagari_jax.md [hunk 2]
```diff
@@ -0,0 +1,925 @@
...
+
+<a href="https://colab.research.google.com/gist/shlff/5bde3e216c5f65dac9763eb17c549044/aiyagari_jax.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>
+
++++ {"id": "b205a26b"}
```
Comment (jstac) [inline lectures/aiyagari_jax.md:18]: same here

File: lectures/aiyagari_jax.md [hunk 3]
```diff
@@ -0,0 +1,925 @@
...
+# The Aiyagari Model
+
+```{code-cell}
+:id: 46444d67
+:incorrectly_encoded_metadata: hide-output=false
+
+#!pip install --upgrade quantecon  #  Uncomment if necessary
+```
+
++++ {"id": "67cab042"}
```
Comment (jstac) [inline lectures/aiyagari_jax.md:29]: same here, etc.

File: lectures/aiyagari_jax.md (hunk 1 above)
Comment (shlff) [reply lectures/aiyagari_jax.md]: Thanks for your nice comments @jstac . 

The commit below fixes these issues and updates the correct ``math-label`` environment for math equations.

**Comment by jstac** [comment]
Thanks guys, please let me know when it's ready for review.

**Comment by Smit-create** [comment]
> Exercise 2 asks the reader to change agent's utility function

I am trying to work on exercise 2 and since it is asking to change the utility function, we will need to re-write all the functions `B`, `T_σ`, `get_value`, etc. I'm not able to write a generalized version of these functions because of:
1. If we declare a global function `u` and change that to some other function won't work -- JAX compiles with global variables and will not change.
2. If we pass the callback then that throws error currently because of unsupported argtype.

So do we want to repeat all those 6-7 functions with a new version just by changing the utility function?

**Comment by jstac** [comment]
Hi @Smit-create , thanks for looking into this.

To change the utility function, you only need to alter `B` and `compute_r_σ`.

In fact, with a little bit of effort, `compute_r_σ` could be obtained from `B`.  Mathematically,

$$ r_\sigma(x) = B(x, \sigma(x), v) $$

So I think the job is not too hard.

**Comment by Smit-create** [comment]
> To change the utility function, you only need to alter `B` and `compute_r_σ`.

I tried but we have the same problem as in the https://github.com/QuantEcon/lecture-jax/pull/19#issuecomment-1514475632 point 1.

See if we use the below script:
```py
@jax.jit
def some_B(x):
    return x

@jax.jit
def someone_using_B():
    x = 5
    return some_B(x)

print(someone_using_B())

@jax.jit
def some_B(x):
    return 2*x

print(someone_using_B())
```

This won't work as `someone_using_B` is` jax.jit` and is not re-compiled, and the output will be:
```
5
5
```

This only works if we have pure Python functions like the following:
```py
def some_A(x):
    return x

def someone_using_A():
    x = 5
    return some_A(x)

print(someone_using_A())

def some_A(x):
    return 2*x

print(someone_using_A())
```
This will print:
```
5
10
```

In order to fix snippet 1, we will need something like this:
```py
@jax.jit
def some_B(x):
    return x

@jax.jit
def someone_using_B():
    x = 5
    return some_B(x)

print(someone_using_B())

@jax.jit
def some_B(x):
    return 2*x


print(someone_using_B())

someone_using_B = jax.jit(someone_using_B)
print(someone_using_B())
```

This will print:
```
5
5
10
```
I am sorry @jstac if I misunderstood your hint. Please feel free to correct me.

**Comment by Smit-create** [comment]
I added exercise 2 using the re-compilation trick, which works now. Please have a look and let me know if anything needs fixing or adding.

Thank you @jstac

**Comment by jstac** [comment]
Great work @Smit-create --- this worked really nicely!

I'm going to push some small edits to this branch.

**Comment by Smit-create** [comment]
Thanks @jstac!
